README.md
# Anomalous Behaviour Forecast

Module responsable for forecasting anomalies occurences based on images from Google Earth Engine API and machine learning



### Dependencies

- Python 3.7.7 64-bit ou superior
- Modules: oauth2client earthengine-api matplotlib pandas numpy requests pillow geojson argparse joblib scipy seaborn sklearn tensorflow



### Attention, before running:

Before running the script and after installing the libraries, you must authenticate with the Google Earth Engine API using the following command:

```
earthengine authenticate
```



### How to execute the default script?

python /path/to/abf/script.py --lat_lon=-83.48811946836814,41.85776095627803,-83.18290554014548,41.677617395337826 --name=erie --from_date=2019-07-11 --reducer --class_mode --disable_attribute_lat_lon




### What are the results?

The script will generate a 5-days forecast of occurrence of algae blooming in the inserted study area starting from input date. Therefore, a folder located in 'data' is created and named based on the date and version of the script executed. Example: /path/to/abf/data/20201026_090433[v=V20-erie,d=2019-07-11,dt=180,din=5,dout=5,m=rf,g=12]. 

ATTENTION: big date range tends to lead to memory leak, stopping the script execution. It is always a good pratice to split the dates in two or three parts, unless you have a big amount of memory in your computer.

The following results are generated:

- timeseries.csv (Time series from dates ranging current day and days threshold parameter)
- results.csv (Accuracy results from the forecasted period validated using indices thresholding approach)
- prediction/csv (CSVs from generated outputs and predictions separated by day)
- prediction/geojson (GeoJSONs from generated outputs and predictions separated by day)
- prediction/image  (Plot from generated outputs and predictions separated by day)



### Example

```
# Import
import ee
from modules import abf

# Initialize Google Earth Engine
ee.Initialize()

# Folder where to save results
folder = "/path/to/desired/folder"

# Create algorithm object
algorithm = abf.Abf(days_threshold=180,
                    grid_size=12,
                    sensor="modis", 
                    geometry=geometry,
                    lat_lon="-83.48811946836814,41.85776095627803,-83.18290554014548,41.677617395337826",
                    cache_path=folder,
                    force_cache=False,
                    days_in=5,
                    days_out=5,
                    from_date="2019-07-11",
                    model="rf",
                    reducer=True,
                    class_mode=True,
                    attribute_lat_lon=False)

# preprocessing
algorithm.process_timeseries_data()
algorithm.process_training_data(df=algorithm.df_timeseries)

# train/predict
algorithm.train(batch_size=2048, disable_gpu=True)
algorithm.predict(folder=folder+"/prediction")

# prediction results
algorithm.save_dataset(df=algorithm.df_results, path=folder+'/results.csv')

# preprocessing results
algorithm.save_dataset(df=algorithm.df_timeseries, path=folder+'/timeseries.csv')
```


CHANGELOG
#####################################################################################################################################
# ### ABF - Anomalous Behaviour Forecast
# ### Script responsible for executing the anomalous behaviour forecast using machine learning and deep learning
# ### Python 3.7 64Bits is required!
#####################################################################################################################################

### Change History
Version 1: 
- Repository creation

Version 2: 
- Accuracy metrics insertion and spreadsheet creation with final results

Version 3: 
- Created spare function to assess training parameters

Version 4: 
- Changed the metrics to evaluate the training process

Version 5: 
- pre-training and results plots fixes

Version 6: 
- model parameters fixes

Version 7: 
- fix in the evaluation process

Version 8: 
- trying to fix de accuracy problem in training process

Version 9: 
- few algorithm fixes

Version 10: 
- algorithm modified to use regression instead of classification (it will predict ndvi, fai and slope)

- Version 11: 
- Multilayer Perceptron and LSTM fixes

Version 12: 
- LSTM to encoding and decoding version, Random Forest and SVM RandomizedSearchCV version, removed mlp_relu and grid fixes
- New MLP and LSTM parametrization
- Module GEE and Misc updateds

Version 13: 
- Added new indices: SABI and NDWI
- Added RandomizedSearchCV to MLP

Version 14: 
- Added parameter 'normalized': when activated, it will normalized indices values to range -1 and 1 (ndwi, ndvi and sabi)
- Added 'class_mode' (regression modelling process) and 'class_weight' (it will assign weights to classes before the training process) parameters
- Added parameter 'propagate': when activated, it will propagate predictions from one day to another in prediction date range
- Added parameter 'rs_train_size', 'rs_iter': it allows increase the randomized search training size and iterations
- Added parameter 'pca_size'

Version 15: 
- Changed accuracy calculation process (removed indetermined bloom pixels)

Version 16:
- Changed some default parameters based on previous tests

Version 17:
- Added parameters for new tests and modification of the LSTM modelling process
- Changed training labels selection, using only edges values (0 and total indexes)

Version 18:
- Fixes in the modelling process and label selection
- Added ROI validation from Google Earth Engine Console

Version 19:
- Fix SVM slow modelling process

Version 20:
- Fix error when using attributes disable_attribute_lat_lon and disable_attribute_doy

Version 21:
- Selection of best parameters and change in scene results format

Version 22:
- Fixes in MLP and LSTM modelling and prediction process
- Changes in grid and scene results

Version 23:
- Fixes in scene plot results

Version 24:
- Fixes in the interpolation process of classification/prediction dataframe

Version 25:
- Fixes in indice_threshold selection
- Changed accuracy calculation method in grid-wise
- Added median, mode and mean final results comparison

Version 26
- Fixes in the grid calculation process
- Fixes in the median calculation results

Version 27
- Fixes in the script.py attributes
- Fixes in the reduction plot
- New batch script (added Lake Taihu area)

Version 28
- Added reference validation (Earth Earth Engine)
- Added reduction parameter (meadin or min, if using referente validation)
- Major fixes in the statistics functions

Version 29
- Changed MNDWI threshold to 0.0

Version 30
- Fixed errors in LSTM modelling process

Version 31
- Fixed erros in results saving process

Version 32
- Added support for shapefile cropping

Version 33
- Changed parameters named to correlate with written paper

script.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

#####################################################################################################################################
# ### ABF - Anomalous Behaviour Forecast
# ### Script responsible for executing the anomalous behaviour forecast using machine learning and deep learning
# ### Python 3.7 64Bits is required!
#####################################################################################################################################

# ### Version
version = "V33"


# ### Module imports

# Main
import ee
import pandas as pd
import math
import requests
import time
import warnings
import os
import sys
import argparse
import traceback
import gc

# Sub
from datetime import datetime as dt
from datetime import timedelta as td

# Extras modules
from modules import misc, gee, abf


# ### Script args parsing

# starting arg parser
parser = argparse.ArgumentParser(description=version)

# create arguments
parser.add_argument('--lat_lon', dest='lat_lon', action='store', default="-83.48811946836814,41.85776095627803,-83.18290554014548,41.677617395337826",
                   help="Two diagnal points (Latitude 1, Longitude 1, Latitude 2, Longitude 2) of the study area")
parser.add_argument('--instant', dest='instant', action='store', default="2013-10-11",
                   help="Date to end time series (it will forecast 5 days starting from this date)")
parser.add_argument('--name', dest='name', action='store', default="erie",
                   help="Place where to save generated files")
parser.add_argument('--spanning_period', dest='spanning_period', action='store', type=int, default=180,
                   help="Spanning period used to build the timeseries and training set")
parser.add_argument('--past_steps', dest='past_steps', action='store', type=int, default=4,
                   help="Past steps to be used as input forecast")
parser.add_argument('--forecasting_period', dest='forecasting_period', action='store', type=int, default=5,
                   help="Forecasting period to be used as output forecast")
parser.add_argument('--spatial_context_size', dest='spatial_context_size', action='store', type=int, default=7,
                   help="Spatial context size in pixels that will be used in continuous/occurrences results")
parser.add_argument('--model', dest='model', action='store', default="rf",
                   help="Select the desired module: mlp, lstm, rf, svm or all (None)")
parser.add_argument('--fill_missing', dest='fill_missing', action='store', default="time",
                   help="Defines algorithm to be used to fill empty dates and values: dummy, ffill, bfill, time, linear")
parser.add_argument('--remove_dummies', dest='remove_dummies', action='store_true',
                   help="Defines if the dummies will be removed before training (only works with fill_missing=dummy)")
parser.add_argument('--scaler', dest='scaler', action='store', default="minmax",
                   help="Defines algorithm to be used in the data scalling process: robust, minmax or standard")
parser.add_argument('--disable_reducer', dest='disable_reducer', action='store_false',
                   help="Defines if reducer will not be applied to remove unnecessary features")
parser.add_argument('--disable_normalization', dest='disable_normalization', action='store_false',
                   help="Defines if normalization (-1,1) will not be applied to indices mndwi, ndvi and sabi")
parser.add_argument('--regression_mode', dest='regression_mode', action='store_false',
                   help="Defines whether will use raw values or classes in the regression models")
parser.add_argument('--class_weight', dest='class_weight', action='store_true',
                   help="Defines whether classes will have defined weights for each")
parser.add_argument('--propagate', dest='propagate', action='store_true',
                   help="Defines whether predictions will be propagated ahead")
parser.add_argument('--rs_train_size', dest='rs_train_size', action='store', type=float, default=500.0,
                   help="It allow increase the randomized search dataset training size (it can be a floater or integer)")
parser.add_argument('--rs_iter', dest='rs_iter', action='store', type=int, default=500,
                   help="It allow increase the randomized search iteration size")
parser.add_argument('--pca_size', dest='pca_size', action='store', type=float, default=0.900,
                   help="Define PCA reducer variance size")
parser.add_argument('--reduction', dest='reduction', action='store', type=str, default="median",
                   help="Define which reduction (median or min) will be used in the reduction stats")
parser.add_argument('--convolve', dest='convolve', action='store_true',
                   help="Define if a convolution box-car low-pass filter will be applied to images before training")
parser.add_argument('--convolve_radius', dest='convolve_radius', action='store', type=int, default=1,
                   help="Define the amont of radius will be used in the convolution box-car low-pass filter")
parser.add_argument('--enable_attribute_lat_lon', dest='enable_attribute_lat_lon', action='store_true',
                   help="Enable attributes lat and lons in the training process")
parser.add_argument('--disable_attribute_doy', dest='disable_attribute_doy', action='store_false',
                   help="Disable attribute doy from training modeling")
parser.add_argument('--disable_shuffle', dest='disable_shuffle', action='store_false',
                   help="Disable data shutffle before splitting into train and test matrix")
parser.add_argument('--save_pairplots', dest='save_pairplots', action='store_true',
                   help="Save pairplots from attributes and indices")
parser.add_argument('--save_train', dest='save_train', action='store_true',
                   help="Enable saving the training dataset (csv)")
parser.add_argument('--cloud_threshold', dest='cloud_threshold', action='store', type=float, default=0.50,
                   help="Define which cloud threshold will be used in the timeseries modelling process")
parser.add_argument('--shapefile', dest='shapefile', action='store',
                   help="Use a shapefile to clip a region of interest")

# parsing arguments
args = parser.parse_args()



# ### Start

try:

  # Start script time counter
  start_time = time.time()

  # Google Earth Engine API initialization
  ee.Initialize()



  # ### Working directory

  # Data path
  folderRoot = os.path.dirname(os.path.realpath(__file__))+'/data'
  if not os.path.exists(folderRoot):
    os.mkdir(folderRoot)

  # Images path
  folderCache = os.path.dirname(os.path.realpath(__file__))+'/cache'
  if not os.path.exists(folderCache):
    os.mkdir(folderCache)



  # ### Selection of coordinates (colon, lat and lon separated by comma, all together) and dates by the user (two dates, beginning and end, separated by commas)
  x1,y1,x2,y2 = args.lat_lon.split(",")

  # Assemble Geometry on Google Earth Engine
  geometry = ee.Geometry.Polygon(
          [[[float(x1),float(y2)],
            [float(x2),float(y2)],
            [float(x2),float(y1)],
            [float(x1),float(y1)],
            [float(x1),float(y2)]]])


  
  # ### ABF execution

  # folder to save results from algorithm at
  folder = folderRoot+'/'+dt.now().strftime("%Y%m%d_%H%M%S")+'[v='+str(version)+'-'+str(args.name)+',m='+str(args.model)+',d='+str(args.instant)+',ct='+str(args.cloud_threshold)+']'
  if not os.path.exists(folder):
    os.mkdir(folder)

  # create algorithm
  algorithm = abf.Abf(days_threshold=args.spanning_period,
                      grid_size=args.spatial_context_size,
                      sensor="modis",
                      geometry=geometry,
                      lat_lon=args.lat_lon,
                      path=folder,
                      cache_path=folderCache,
                      force_cache=False,
                      morph_op=None, 
                      morph_op_iters=1,
                      convolve=args.convolve,
                      convolve_radius=args.convolve_radius,
                      scaler=args.scaler,
                      days_in=args.past_steps,
                      days_out=args.forecasting_period,
                      from_date=args.instant,
                      model=args.model,
                      fill_missing=args.fill_missing,
                      remove_dummies=args.remove_dummies,
                      reducer=args.disable_reducer,
                      normalized=args.disable_normalization,
                      class_mode=args.regression_mode,
                      class_weight=args.class_weight,
                      propagate=args.propagate,
                      rs_train_size=args.rs_train_size,
                      rs_iter=args.rs_iter,
                      pca_size=args.pca_size,
                      attribute_lat_lon=args.enable_attribute_lat_lon,
                      attribute_doy=args.disable_attribute_doy,
                      shuffle=args.disable_shuffle,
                      test_mode=False,
                      shapefile=args.shapefile,
                      cloud_threshold=args.cloud_threshold)

  # preprocessing
  algorithm.process_timeseries_data()
  algorithm.process_training_data(df=algorithm.df_timeseries)

  # save pairplots
  if args.save_pairplots == True:
    if not os.path.exists(folder+"/pairplots"):
      os.mkdir(folder+"/pairplots")
    algorithm.save_attributes_pairplot(df=algorithm.df_timeseries,path=folder+"/pairplots/attributes.png")
    algorithm.save_indices_pairplot(df=algorithm.df_timeseries,path=folder+"/pairplots/indices.png")
    
  # train/predict
  else:
    algorithm.train(batch_size=2048, disable_gpu=True)
    algorithm.predict(folder=folder+"/prediction", path='users/pedroananias/'+str(args.name), chla_threshold=20)
    algorithm.predict_reduction(folder=folder+"/prediction", reduction=str(args.reduction))

    # prediction results
    algorithm.save_dataset(df=algorithm.df_results, path=folder+'/results.csv')
    algorithm.save_dataset(df=algorithm.df_scene, path=folder+'/results_scene.csv')

    # preprocessing results
    algorithm.save_dataset(df=algorithm.df_timeseries, path=folder+'/timeseries.csv')

    # save training datasets
    if args.save_train == True:
      algorithm.save_dataset(df=pd.DataFrame(algorithm.df_train[0]), path=folder+'/df_train_X.csv')
      algorithm.save_dataset(df=pd.DataFrame(algorithm.df_train[1]), path=folder+'/df_train_y.csv')
      algorithm.save_dataset(df=pd.DataFrame(algorithm.df_test[0]), path=folder+'/df_test_X.csv')
      algorithm.save_dataset(df=pd.DataFrame(algorithm.df_test[1]), path=folder+'/df_test_y.csv')
      algorithm.save_dataset(df=pd.DataFrame(algorithm.df_classification), path=folder+'/df_classification_X.csv')

  # results
  # add results and save it on disk
  path_df_results = folderRoot+'/results.csv'
  df_results = pd.read_csv(path_df_results).drop(['Unnamed: 0'], axis=1, errors="ignore").append(algorithm.df_results) if os.path.exists(path_df_results) else algorithm.df_results.copy(deep=True)
  df_results.to_csv(r''+path_df_results)

  # clear memory
  del algorithm
  gc.collect()

  # ### Script termination notice
  script_time_all = time.time() - start_time
  debug = "***** Script execution completed successfully (-- %s seconds --) *****" %(script_time_all)
  print()
  print(debug)

except:

    # ### Script execution error warning

    # Execution
    print()
    print()
    debug = "***** Error on script execution: "+str(traceback.format_exc())
    print(debug)

    # Removes the folder created initially with the result of execution
    script_time_all = time.time() - start_time
    debug = "***** Script execution could not be completed (-- %s seconds --) *****" %(script_time_all)
    print(debug)
